content/posts/codegen.md

```diff
@@ -0,0 +1,143 @@
+in Google's build system.
+
+Quick preface: I'm talking about writing a computer program that outputs
+some kind of file that's funny designed to be read in by a compiler or
+interpreter. I'm not talking about writing data files or schema files. It feels
+natural for computers to be writing those, since they broadly match up with the
+native data types found in programming languages.
+
+## Why Generate Code?
+
+My team at Google is responsible for building open source support for all of
+Google Cloud's different resource types for various declarative tooling like
+[Terraform](https://github.com/hashicorp/terraform-provider-google) or
+[Ansible](https://github.com/ansible-collections/google.cloud). We've even
+open-sourced one of our [code
+generators](https://github.com/GoogleCloudPlatform/magic-modules) if you'd like
+to take a look!
+
+If you haven't seen GCP lately, there's an unbelievable amount of resource types
+(in the order of hundreds!). Everytime a user wishes to work with a resource of
+a certain type in one of those tools, there has to be code in that cool that
+handles that particular resource type.
+
+Writing support for each of these resource types looks pretty
+similar. They all follow the same rough logic:
+
+- A user needs a cloud resource to exist in a certain manner
+- The tool calls that cloud resource's GET API to see if the resource exists.
+- If it doesn't, the tool calls the resource type's Create API.
+- If the resource does exist, the tool checks to see if the user's preferences
+  differ from whatever the resource currently has.
+- If it does, the update API gets called.
+
+You might first think that we need a function to handle this. The function could
+take in the user's intention and links to Get/Create/Update APIs. But, as time
+goes on, you realize that some resources handle the creation process slightly
+different from others. Then, every resource type expresses their current state
+(from the GET API) slightly differently from each other. That makes handling the
+"check if user's preferences are different" phase slightly different for each
+resource. As you go on, you realize that parts of each resource type's code will
+be identical to all the others and other parts will be radically different.
+
+Each resource type also needs to have a file with a handful of boilerplate
+methods defined. No amount of functional or class abstraction can change the
+fact that our team needs to write out all of those boilerplate methods...and
+that's before the fact that we realize that the boilerplate isn't just
+copy-paste.
+
+Doing this once or twice is interesting! Doing it a dozen or so times feels
+monotonous, especially once you realize that you made the same bug in all dozen
+resource types and have to make the same fix twelve times. Now, imagine doing
+it a hundred times.
+
+Code generation lets us create a source-of-truth for all this boilerplate.
```

Comment:
Really well written section overall -- it almost feels like you're telling a story, the way you're narratively walking the reader through the motivation here.